go/libkb/merkle_tools.go

```diff
@@ -5,6 +5,103 @@ import (
+// and returns the corresponding object.
+func leafPart(leaf *MerkleGenericLeaf, typ keybase1.SeqType) *MerkleTriple {
+	if leaf == nil {
+		return nil
+	}
+	switch typ {
+	case keybase1.SeqType_PUBLIC:
+		return leaf.Public
+	case keybase1.SeqType_SEMIPRIVATE:
+		return leaf.Private
+	default:
+		return nil
+	}
+}
+
+// lookupLeafAtRootSeqno queries the server for the merkle root at rootSeqno, and then descends the tree to the
+// leaf for the supplied id. It returns the leaf, the root, and the leaf part determed by SeqType.
+func lookupLeafAtRootSeqno(m MetaContext, id keybase1.UserOrTeamID, rootSeqno keybase1.Seqno, typ keybase1.SeqType) (leaf *MerkleGenericLeaf, root *MerkleRoot, userChainSeqno keybase1.Seqno, err error) {
+	m.CDebugf("lookupLeafAtRootSeqno(%s,%d)", id, rootSeqno)
+	leaf, root, err = m.G().GetMerkleClient().LookupLeafAtSeqno(m.Ctx(), id, rootSeqno)
+	if err != nil {
+		return nil, nil, 0, err
+	}
+	part := leafPart(leaf, typ)
+	if part == nil {
+		return nil, nil, 0, MerkleClientError{fmt.Sprintf("leaf at root %d returned with nil %v part", rootSeqno, typ), merkleErrorBadLeaf}
+	}
+	return leaf, root, part.Seqno, nil
+}
+
+// findFirstLeafWithChainSeqno finds the first chronological leaf in the merkle tree that contains the given
+// chainSeqno for the given UID. Start this search from the given prevRootSeqno, not from merkle position 1.
+// The algorithm is to jump forward, in exponentially larger jumps, until we find a root that has a leaf
+// that is >= the given chainSeqno. Then, to binary search to find the exact [a,b] pair in which a doesn't
+// contain the leaf seqno, and b does. The leaf and root that correspond to be are returned.
+func findFirstLeafWithChainSeqno(m MetaContext, id keybase1.UserOrTeamID, chainSeqno keybase1.Seqno, typ keybase1.SeqType, prevRootSeqno keybase1.Seqno) (leaf *MerkleGenericLeaf, root *MerkleRoot, err error) {
+	defer m.CTrace(fmt.Sprintf("findFirstLeafWithChainSeqno(%s,%d,%d)", id, chainSeqno, prevRootSeqno), func() error { return err })()
+
+	if m.G().Env.GetRunMode() == ProductionRunMode && prevRootSeqno < FirstProdMerkleTreeWithModernShape {
+		return nil, nil, MerkleClientError{"can't operate on old merkle sequence number", merkleErrorOldTree}
+	}
+
+	cli := m.G().GetMerkleClient()
+	low := prevRootSeqno
+	lastp := cli.LastRoot().Seqno()
+	if lastp == nil {
+		return nil, nil, MerkleClientError{"unexpected nil high seqno", merkleErrorBadRoot}
+	}
+	last := *lastp
+	var hi keybase1.Seqno
+	inc := keybase1.Seqno(1)
+
+	// First bump the hi pointer up to a merkle root that overshoots (or is equal to)
+	// the request chainSeqno. Don't go any higher than the last known Merkle seqno.
+	for hi = low + 1; hi < last; hi += inc {
+		var tmpSeqno keybase1.Seqno
+		leaf, root, tmpSeqno, err = lookupLeafAtRootSeqno(m, id, hi, typ)
+		if err != nil {
+			return nil, nil, err
+		}
+		if tmpSeqno >= chainSeqno {
+			break
+		}
+		inc *= 2
+	}
+	if hi > last {
```

Comment:
in case we come here because we overshoot `last` seqno in previous loop, should we check if chainSeqno is lower than last?

Reply:
it's worth a check, it should never happen though.... ok

Reply:
I guess it'll then fail the binary search, which is fine, just wastes some cycles

Reply:
added an error...